### ISSUE TITLE
Using set in protocol supertype seems to break compatibility

### ISSUE THREAD

**Issue by Why-not-now** [python/mypy#15679]
**Bug Report**

When using set, (tuple and list doesn't show an error) using this formats seem to break compatibility.

**To Reproduce**

```python
from __future__ import annotations

from typing import Protocol


class MyProtocol(Protocol):
    def my_method(self) -> set[MyProtocol]:
        ...


class MyClass(MyProtocol):
    def my_method(self) -> set[MyClass]:
        return {self}

```
**Actual Behavior**

`Return type "set[MyClass]" of "my_method" incompatible with return type "set[MyProtocol]" in supertype "MyProtocol"  [override]`

**Your Environment**

- Mypy version used: mypy 1.4.1
- Python version used:

**Comment by hauntsaninja** [comment]
list errors as well. This is expected because list and set are invariant. See https://mypy.readthedocs.io/en/stable/generics.html#variance-of-generic-types
